### PR TITLE
charmcraft: init at 2.6.0

### PR DESCRIPTION
**Pull request by jnsgruk**
## Description of changes

Follow on from #295646 to add the `craft` build tools to NixOS.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/development/python-modules/macaroonbakery/default.nix [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+  setuptools,
+  httmock,
+  fixtures,
+  pytestCheckHook,
+  mock,
+}:
+
+buildPythonPackage rec {
+  pname = "macaroonbakery";
+  version = "1.3.4";
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/macaroonbakery/default.nix:20]: Can we set pyproject = true?

File: pkgs/by-name/ch/charmcraft/package.nix [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+      flake8
+      pydocstyle
+      pyfakefs
+      pytest-check
+      pytest-mock
+      pytest-subprocess
+      pytestCheckHook
+      responses
+    ]
+    ++ [ git ];
```
Comment (SuperSandro2000) [inline pkgs/by-name/ch/charmcraft/package.nix:56]:
```suggestion
  nativeCheckInputs = with python3Packages; [
    pyfakefs
    pytest-check
    pytest-mock
    pytest-subprocess
    pytestCheckHook
    responses
  ] ++ [ git ];
```
we don't want to verify the code quality, as it only creates chores for us and doesn't really prove anything

File: pkgs/development/python-modules/craft-providers/default.nix [hunk 3]
```diff
@@ -46,7 +46,7 @@ buildPythonPackage rec {
     # The urllib3 incompat: https://github.com/msabramo/requests-unixsocket/pull/69
     # This is already patched in nixpkgs.
     substituteInPlace pyproject.toml \
-      --replace-fail "setuptools==67.8.0" "setuptools" \
+      --replace-fail "setuptools==69.1.1" "setuptools" \
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/craft-providers/default.nix:49]: Why not use the relaxDeps hook?

Comment (jnsgruk) [reply pkgs/development/python-modules/craft-providers/default.nix:49]: This doesn't seem to work with pyproject-only projects, or at least I can't make it work

**Comment by jnsgruk** [comment]
@SuperSandro2000 thank you! Comments addressed!

**Comment by jnsgruk** [comment]
@SuperSandro2000 I also snuck in an extra commit to remove the `nix-update-script` from `pydantic-yaml-0`, as I keep getting automated PRs to update it beyond version 0: https://github.com/NixOS/nixpkgs/pull/303678#event-12453183224

**Review by drupol — APPROVED** [review]

**Review by drupol — CHANGES_REQUESTED** [review]
Oops, looks like the CI is broken at https://github.com/NixOS/nixpkgs/pull/303369/checks?check_run_id=23875313079

File: pkgs/development/python-modules/craft-parts/default.nix [hunk 4]
```diff
@@ -1,26 +1,28 @@
...
-, pytest-check
-, pytest-mock
-, pytest-subprocess
-, requests-mock
-, hypothesis
-, git
-, squashfsTools
-, setuptools
-, setuptools-scm
+{
```
[the remark this answers is not in the document]
Comment (jnsgruk) [reply pkgs/development/python-modules/craft-parts/default.nix]: It was - reverted for now and just added the test changes.

[the remark this answers is not in the document]
Comment (jnsgruk) [reply pkgs/development/python-modules/craft-parts/default.nix]: Yep, that should now be the case :)

**Comment by jnsgruk** [comment]
@drupol I think the tests problem is fixed :)

**Review by drupol — APPROVED** [review]

**Comment by jnsgruk** [comment]
@drupol thank you! much appreciated! :rocket:

**Pull request merged**